src/main/java/com/mongodb/hibernate/internal/translate/AbstractMqlTranslator.java

```diff
@@ -582,6 +613,48 @@ private AstSortField createAstSortField(Expression sortExpression, AstSortOrder
         return new AstSortField(fieldPath, astSortOrder);
     }
 
+    // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    // LIMIT/OFFSET/FETCH clause
+
+    @Override
+    public void visitOffsetFetchClause(QueryPart queryPart) {
+        var stages = createSkipAndLimitStages(queryPart);
+        astVisitorValueHolder.yield(SKIP_LIMIT_STAGES, stages);
+    }
+
+    private List<AstStage> createSkipAndLimitStages(QueryPart queryPart) {
+        if (queryPart.getFetchClauseType() != ROWS_ONLY) {
+            throw new FeatureNotSupportedException(format(
+                    "%s does not support '%s' fetch clause type", MONGO_DBMS_NAME, queryPart.getFetchClauseType()));
+        }
+        var stages = new ArrayList<AstStage>(2);
+        final Expression skipExpression;
+        final Expression limitExpression;
+        if (queryPart.isRoot() && limit != null && !limit.isEmpty()) {
```

Comment:
`QueryOptions`'s `Limit` only applies for top-level query (or not in subquery whose `isRoot()` would return false).